frame/contracts/src/lib.rs

```diff
@@ -574,18 +574,21 @@ pub mod pallet {
 			data: Vec<u8>,
 		) -> DispatchResultWithPostInfo {
 			let gas_limit: Weight = gas_limit.into();
-			let origin = ensure_signed(origin)?;
```

Comment:
So by leaving it this way for all other dispatchables, we "forbid" calling them by root, right?

If so, shouldn't we map the error (in all other dispatchables) to `RootNotAllowed` (instead of `BadOrigin`), once having this DispatchError added by this very PR?

Reply:
[`ensure_signed_or_root`](https://github.com/paritytech/substrate/blob/master/frame/system/src/lib.rs#L936) does actually allow 'Root'. What we are preventing here is to let origins that are not signed nor root

Reply:
What I mean here is that in other dispatchables we still use `ensure_root`: 

https://github.com/paritytech/substrate/pull/13708/files#diff-70e9723e9db62816e35f6f885b6770a8449c75a6c2733e9fa7a245fe52c4656cR642-R651

It [returns](https://paritytech.github.io/substrate/master/frame_system/fn.ensure_root.html) `BadOrigin` when it's Root, right? Wouldn't it be better to explicitly handle the `Origin::Root` case in such places and return `RootNotAllowed` when it's called by Root?

Reply:
Yes, `ensure_signed` returns `BadOrigin` when it's root, but it also returns `BadOrigin` for other origins that are not `Signed`. Bear in mind that `ensure_signed` takes a `RuntimeOrigin` enum as parameter (which is not the same as our 'contract' `Origin`) that could have origins other than `Signed` and `Root`

Reply:
So once we want to forbid the Root origin for some dispatchables, it could make sense to use ensure_signed_or_root in all those dispatchables to handle the Root case explicitly and return RootNotAllowed instead of BadOrigin. Also I think it worth moving that origin checks logic into the trait, i.e. Invokable::ensure_origin() method.

Reply:
We don't forbid `RuntimeOrigin::Root` for some dispatchables, we forbid anything that is not 'RuntimeOrigin::Signed()`. Then `BadOrigin` is a generic response for the `ensure_signed` that takes the enum `RuntimeOrigin`. We don't know all the possibilities of that enum, as we don't control it. Also, if we want to forbid the special case `RuntimeOrigin::Root` explicitly, why not forbidding all the other special cases explicitly too?

On the other hand `Origin::Root` belongs to the `Origin` enum, that is part of the pallet contracts and an enum that we control and we know all of it possibilities. Therefore we can return a `RootNotAllowed` cause we know all the other options are actually allowed

Reply:
In the latest commit this logic has been split into two. First we try to transform from `RuntimeOrigin` to `Origin`, if that's not possible a `BadOrigin` is returned.
Then each invokable is responsible for determine whether the given `Origin` is valid or not. In the particular case of the instantiations, if a `Origin::Root` is received it returns `RootNotAllowed`